File: pkg/kernel/linux/lkm/hello.rs
// SPDX-License-Identifier: GPL-2.0

#![no_std]
#![feature(allocator_api, global_asm)]

use kernel::prelude::*;

module! {
  type: Hello,
  name: "hello",
  author: "Schubert Anselme <schubert@anselm.es",
  description: "Linux Kernel Module",
  license: "GPL v2",
  params: {},
}

struct Hello;

impl KernelModule for HelloWorld {
  fn init() -> KernelResult<Self> {
    pr_info("Hello world, from the kernel!\n");
    Ok(Hello {})
  }
}

impl Drop for Hello {
  fn drop(&mut self) {
    pr_info!("Goodby!\n");
  }
}

File: pkg/kernel/linux/lkm/legacy/hello.c
// SPDX-License-Identifier: GPL-2.0

#include <linux/module.h>
#include <linux/init.h>

MODULE_AUTHOR("Schubert Anselme");
MODULE_DESCRIPTION("Linux Kernel Module");
MODULE_LICENSE("GPL v2");

static int __init load(void)
{
  printk("hello - Hello, Kernel!\n");
  return 0;
}

static void __exit unload(void)
{
  printk("hello - Goodby, Kernel\n");
}

module_init(load);
module_exit(unload);
